module/os/linux/zfs/zpl_super.c

```diff
@@ -102,6 +105,40 @@ zpl_evict_inode(struct inode *ip)
 	spl_fstrans_unmark(cookie);
 }
 
+static void
+zpl_umount_begin(struct super_block *sb)
+{
+	zfsvfs_t *zfsvfs = sb->s_fs_info;
+
+	if (zfsvfs) {
+		/*
+		 * Flush out all POSIX I/Os.  Notify all waiters that they
+		 * must end, then wait for all users to drop their holds on
+		 * z_teardown_*_lock, and evict buffers.
+		 */
+		zfsvfs->z_unmounted = Z_UNMOUNT_BEGIN;
+		(void) dmu_objset_shutdown_register(zfsvfs->z_os);
+		rrm_enter(&zfsvfs->z_teardown_lock, RW_WRITER, FTAG);
+		rw_enter(&zfsvfs->z_teardown_inactive_lock, RW_WRITER);
+		rw_exit(&zfsvfs->z_teardown_inactive_lock);
+		rrm_exit(&zfsvfs->z_teardown_lock, FTAG);
+		dmu_objset_evict_dbufs(zfsvfs->z_os);
+
+		dsl_dir_cancel_waiters(zfsvfs->z_os->os_dsl_dataset->ds_dir);
+		dmu_objset_shutdown_unregister(zfsvfs->z_os);
+
+		/*
+		 * It's necessary to clear z_unmounted, unless the user wants
+		 * the filesystem to remain 'unmounted' for the purpose of
+		 * forced pool export.  Retaining this state causes the
+		 * filesystem to reject all but what's required to enable it
+		 * to remain in the mount table.
+		 */
+		if (!zfs_forced_export_unmount_enabled)
+			zfsvfs->z_unmounted = Z_MOUNTED;
```

Comment:
You're right about the race, unlikely though it is.

Regarding using a second boolean instead of a tristate, it's a trade-off that I wasn't sure would be accepted; after all, the extra byte will only be used in rare cases.  Another consequence of using an extra boolean is that most ZPL I/Os would need to read both of them.